File: DSA/LC/98_validate_binary_search_tree.py
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        def rec(root, min_v, max_v):
            if root is None:
                return True

            if root.val >= max_v or root.val <= min_v:
                return False

            return rec(root.left, min_v, root.val) and rec(root.right, root.val, max_v)

        return rec(root, float('-Inf'), float('Inf'))

File: DSA/LC/98_validate_binary_search_tree.ts
class TreeNode<T> {
  value: T;
  left: TreeNode<T> | null = null;
  right: TreeNode<T> | null = null;
  constructor(value: T) {
    this.value = value;
  }
}

function isValidBST(root: TreeNode<number> | null): boolean {
  let prev: number | null = null;

  function inorder(node: TreeNode<number> | null): boolean {
    if (node === null) return true;

    // In-order traversal: left -> node -> right
    if (!inorder(node.left)) return false;

    // Compare current node value with the previous one
    if (prev !== null && node.value <= prev) return false;
    prev = node.value;

    return inorder(node.right);
  }

  return inorder(root);
}

export {};
